packages/contracts/contracts/delegateTargets/VirtualAppAgreement.sol

```diff
@@ -12,47 +12,49 @@ contract VirtualAppAgreement {
   using Transfer for Transfer.Terms;
 
   struct Agreement {
-    Registry registry;
-    NonceRegistry nonceRegistry;
+    address registryAddr;
     Transfer.Terms terms;
+    uint256 expiry;
+    bytes32 target;
     uint256 capitalProvided;
-    uint256 loanDuration;
-    bytes32 virtualStateDeposit;
-    address intermediary;
+    address[2] beneficiaries;
   }
 
   function delegateTarget(
-    bytes32 salt,
-    uint256 nonce,
     Agreement agreement
   )
     public
   {
+    Registry registry = Registry(agreement.registryAddr);
+    require(agreement.expiry <= block.number, "agreement has not expired yet");
+    address target = registry.resolver(agreement.target);
+    Transfer.Transaction memory resolution = AppInstance(target).getResolution();
 
-    require(
-      agreement.nonceRegistry.isFinalized(salt, nonce),
-      "VirtualAapp nonceRegistry was not finalized"
-    );
-
-    address[] memory to = new address[](2);
-    to[0] = agreement.registry.resolver(agreement.virtualStateDeposit);
-    to[1] = agreement.intermediary;
+    require(resolution.value.length == 1, "returned invalid resolution");
+    require(resolution.to.length == 1, "returned invalid resolution");
+    require(agreement.terms.assetType == resolution.assetType, "returned incompatible resolution");
+    require(agreement.terms.token == resolution.token, "returned incompatible resolution");
+    require(agreement.capitalProvided > resolution.value[0], "returned incompatible resolution");
 
     uint256[] memory amount = new uint256[](2);
-    amount[0] = agreement.capitalProvided;
-    amount[1] = agreement.capitalProvided;
+    amount[0] = resolution.value[0];
+    amount[1] = agreement.capitalProvided - amount[0];
+
     bytes[] memory data = new bytes[](2);
```

Comment:
Generally speaking, it's annoying that we have to construct all these arguments every time. Is there a way we can create a `LibTransaction` that has convenience methods for constructing 2-person payouts of ETH for example to clean up this code?

Reply:
Yes, I had a (really) old PR on the legacy contracts repo for this. I can reopen that PR separately from this one